kernel/src/sched/sched_class.rs

```diff
@@ -0,0 +1,235 @@
+// SPDX-License-Identifier: MPL-2.0
+
+#![warn(unused)]
+
+use alloc::{
+    boxed::Box,
+    sync::{Arc, Weak},
+    vec,
+    vec::Vec,
+};
+use core::{any::Any, fmt};
+
+use ostd::{
+    cpu::{num_cpus, CpuSet, PinCurrentCpu},
+    task::{
+        scheduler::{inject_scheduler, EnqueueFlags, LocalRunQueue, Scheduler, UpdateFlags},
+        Priority, Task,
+    },
+    trap::disable_local,
+};
+
+mod time;
+
+mod fair;
+mod idle;
+mod real_time;
+mod stop;
+
+use ostd::arch::read_tsc as sched_clock;
+
+use crate::thread::Thread;
+
+#[allow(unused)]
+pub fn init() {
+    inject_scheduler(Box::leak(Box::new(ClassScheduler::default())));
+}
+
+/// The kind of scheduler classes.
+#[derive(Debug, Clone, Copy)]
+#[repr(u8)]
+pub enum SchedKind {
+    Stop,
+    RealTime,
+    Fair,
+    Idle,
+}
+
+/// Represents the middle layer between scheduling classes and generic scheduler
+/// traits. It consists of a list of run queues for CPU cores. Other global
+/// information may also be stored here.
+pub struct ClassScheduler {
+    rqs: Vec<spin::Mutex<ClassRq>>,
+}
+
+/// Represents the run queue for CPU cores. It stores a list of run queues for
+/// scheduling classes in one CPU. The current task of this CPU core is also
+/// stored in this structure.
+struct ClassRq {
+    class: Vec<Box<dyn SchedClassRq>>,
+    current: Option<Arc<Task>>,
+}
+
+/// The run queue for scheduling classes (the main trait). Scheduling classes
+/// should implement this trait to function as expected.
+trait SchedClassRq: Send {
+    /// The kind of this scheduling class.
+    #[allow(unused)]
+    fn kind(&self) -> SchedKind;
+
+    /// Enqueues a task into the run queue.
+    fn enqueue(&mut self, thread: Arc<Thread>);
+
+    /// Dequeues a task from the run queue.
+    fn dequeue(&mut self, thread: &Arc<Thread>);
+
+    /// Picks the next task for running.
+    fn pick_next(&mut self) -> Option<Arc<Thread>>;
+
+    /// Update the information of the current task.
+    fn update_current(&mut self, cur: &Thread, flags: UpdateFlags) -> bool;
+
+    /// Show debug information about this object.
+    #[allow(unused)]
+    fn debug(&self) -> &dyn fmt::Debug;
+}
+
+/// The scheduling entity. Each scheduling class can cast the trait object to
+/// its own data using `as_any().downcast_ref()`/`as_any_mut().downcast_mut()`.
+pub trait SchedEntity: Any + Send + Sync {
+    /// Shows the kind of its scheduling class.
+    fn kind(&self) -> SchedKind;
+
+    /// Convenience for downcasting.
+    fn as_any(&self) -> &dyn Any;
```

Comment:
This is because methods on trait objects cannot be called by objects of its sub traits if `#[feature(trait_upcasting)]` is not enabled, which I didn't notice in `kernel/lib.rs` and wrote it by habit.

Reply:
So they can and should be removed, right?

Reply:
Yes. I've already removed it and `debug` in the latest commit.